rbd/migration_test.go

```diff
@@ -0,0 +1,174 @@
+//go:build !(octopus || nautilus) && ceph_preview
+// +build !octopus,!nautilus,ceph_preview
+
+package rbd
+
+import (
+	"testing"
+
+	"github.com/ceph/go-ceph/rados"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func createAndWriteDataToImage(t *testing.T, ioctx *rados.IOContext) string {
+	opts := NewRbdImageOptions()
+	defer opts.Destroy()
+
+	assert.NoError(t,
+		opts.SetUint64(ImageOptionOrder, uint64(testImageOrder)))
+
+	name := GetUUID()
+	err := CreateImage(ioctx, name, 1<<25, opts)
+	require.NoError(t, err)
+
+	img, err := OpenImage(ioctx, name, NoSnapshot)
+	require.NoError(t, err)
+
+	img.CreateSnapshot("snap1")
+
+	// write some data
+	_, err = img.WriteAt([]byte("sometimes you feel like a nut"), 0)
+	assert.NoError(t, err)
+
+	err = img.Close()
+	assert.NoError(t, err)
+
+	return name
+}
+
+func TestMigration(t *testing.T) {
+	conn := radosConnect(t)
+
+	pool := GetUUID()
+
+	err := conn.MakePool(pool)
+	assert.NoError(t, err)
+
+	ioctx, err := conn.OpenIOContext(pool)
+	require.NoError(t, err)
+
+	// setup image
+	name := createAndWriteDataToImage(t, ioctx)
+
+	// call MigrationPrepare
+	var opts MigrationOptions
+	opts.SourceImage = name
+	opts.DestImage = GetUUID()
+	err = MigrationPrepare(ioctx, ioctx, opts, NewRbdImageOptions())
+	assert.NoError(t, err)
+
+	// get status
+	status, err := MigrationStatus(ioctx, opts.DestImage)
+	assert.NoError(t, err)
+	assert.Equal(t, status.State, MigrationImagePrepared)
+
+	// call MigrationExecute
+	err = MigrationExecute(ioctx, opts.DestImage)
+	assert.NoError(t, err)
+
+	// get status
+	status, err = MigrationStatus(ioctx, opts.DestImage)
+	assert.NoError(t, err)
+	assert.Equal(t, status.State, MigrationImageExecuted)
+
+	// call MigrationCommit
+	err = MigrationCommit(ioctx, opts.DestImage)
+	assert.NoError(t, err)
+
+	// get status returns error
+	_, err = MigrationStatus(ioctx, opts.DestImage)
+	assert.Error(t, err)
+
+	// open the target image
+	_, err = OpenImage(ioctx, opts.DestImage, NoSnapshot)
+	assert.NoError(t, err)
+
+	// original image is moved to trash as part of migration prepare
+	_, err = OpenImage(ioctx, opts.SourceImage, NoSnapshot)
+	assert.Error(t, err)
+
+	ioctx.Destroy()
+	conn.DeletePool(pool)
+	conn.Shutdown()
+
+}
+
+func TestMigrationPrepareImport(t *testing.T) {
+	conn := radosConnect(t)
+
+	pool := GetUUID()
+	err := conn.MakePool(pool)
+	assert.NoError(t, err)
+
+	ioctx, err := conn.OpenIOContext(pool)
+	require.NoError(t, err)
+
+	// setup image
+	name := createAndWriteDataToImage(t, ioctx)
+
+	// call MigrationPrepareImport
+	var opts MigrationOptions
+	opts.DestImage = GetUUID()
+	opts.SourceSpec = `{"type": "native", "snap_name": "snap1", "pool_name":"` + pool + `", "image_name": "` + name + `"}}`
```

Comment:
Is the JSON structure for a Source Spec documented by Ceph? If so it may be useful to prove a function that generates the JSON string rather than just "hand assembling" it like this test does. If it's more ad-hoc I won't object to the test building the JSON string this way... but I think it's worth considering doing it in a more structured manner.

Reply:
The json spec seems very specific to the user's needs but I could try implementing helper functions for each. I'm thinking something like creating multiple functions such as `CreateSourceSpecS3`, `CreateSourceSpecHttp`, `CreateSourceSpecFile` and `CreateSourceSpecNative`? 


ref: https://docs.ceph.com/en/pacific/rbd/rbd-live-migration/#formats